## Activity 2: K-nn classification with credit data

### Install libraries

In [16]:
library(class)

### Load CSV file and remove NA values

In [17]:
data1 <- read.csv("german_credit_data1.csv");
data1 <- na.omit(data1)
head(data1)

X. Age Sex    Job Housing Saving.accounts Checking.account Credit.amount
2   1 22  female 2   own     little          moderate         5951         
4   3 45  male   2   free    little          little           7882         
5   4 53  male   2   free    little          little           4870         
8   7 35  male   3   rent    little          moderate         6948         
10  9 28  male   3   own     little          moderate         5234         
11 10 25  female 2   rent    little          moderate         1295         
   Duration Purpose             Credit.Risks
2  48       radio/TV            2           
4  42       furniture/equipment 1           
5  24       car                 2           
8  36       car                 1           
10 30       car                 2           
11 12       car                 2

### Factorize the value of categorical data to numerical data

In [18]:
data1$Housing <- as.numeric(factor(data1$Housing))
data1$Saving.accounts <- as.numeric(factor(data1$Saving.accounts))
data1$Checking.account <- as.numeric(factor(data1$Checking.account))
data1$Purpose <- as.numeric(factor(data1$Purpose))

### Calculate KNN samplewise and by changing the k values

In [19]:
sample <- c(0.6,0.7,0.8)
extra = c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20)
for(sampling_rate in sample) {
    #sampling_rate <- 
    data1 <- as.data.frame(data1)
    trainrows <- nrow(data1)*sampling_rate
    train_data <- data1[1:trainrows,]
    test_data <- data1[trainrows:nrow(data1),]
    num.train.rows <- nrow(train_data)
    num.test.rows <- nrow(test_data)
    cltrain <- data1$Credit.Risks[1:trainrows]
    cltest <- data1$Credit.Risks[trainrows:nrow(data1)]
    final_train <- subset(train_data, select = c(Age,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose,Credit.Risks))
    final_test <- subset(test_data, select = c(Age,Job,Housing,Saving.accounts,Checking.account,Credit.amount,Duration,Purpose,Credit.Risks))
    l = list()
    for (k in 1:20) {
        predicted_labels <- knn(final_train, final_test, cl = cltrain, k)
        num.incorrect.labels <- sum(predicted_labels != cltest)
        misclassificaiton_rate <- num.incorrect.labels/num.test.rows
        l <- c(l,misclassificaiton_rate)
    }
    extra <- cbind(extra,l)
}

### Representing data in tabluar form

In [20]:
extra1 <- as.data.frame(extra)
colnames(extra1) <- c("K Values","Sampling Rate 0.6","Sampling Rate 0.7","Sampling Rate 0.8")
extra1

K Values Sampling Rate 0.6 Sampling Rate 0.7 Sampling Rate 0.8
1  1        0.4641148         0.5095541         0.5142857        
2  2        0.4545455         0.5286624         0.5428571        
3  3        0.4258373         0.4394904         0.4666667        
4  4        0.4593301         0.4267516         0.4285714        
5  5        0.3971292         0.4076433         0.4380952        
6  6        0.3875598         0.4522293         0.4761905        
7  7        0.3971292         0.4140127         0.4380952        
8  8        0.3732057         0.433121          0.3904762        
9  9        0.3684211         0.3694268         0.3809524        
10 10       0.3923445         0.388535          0.4              
11 11       0.3827751         0.388535          0.4190476        
12 12       0.3349282         0.3821656         0.3904762        
13 13       0.3492823         0.3757962         0.3714286        
14 14       0.3588517         0.388535          0.4285714        
15 15       0.3732057         0.3630573         0.4190476        
16 16       0.3684211         0.3566879         0.4095238        
17 17       0.3732057         0.343949          0.4095238        
18 18       0.3923445         0.3757962         0.4095238        
19 19       0.3923445         0.3566879         0.4190476        
20 20       0.4019139         0.388535          0.3904762

#### The value of misclassification rate varies from 0.33 to 0.46 while taking 60% training data & 40% testing data, 0.34 to 0.52 while taking 70% training data & 30% testing data and 0.37 to 0.54 in 80% training data & 20% testing data. 